Add kRange preceding/following frames in window fuzzer

pramodsatya opened the pull request:
1. Adds support for kRange preceding/following frames in window fuzzer
2. Adds [reference query runner context](https://ibm.box.com/s/9tuk22hfp13imjwq9xelnz1dogsdgdh1) for PrestoSql frame clause for kRange
preceding/following frames.

Resolves https://github.com/facebookincubator/velox/issues/9572 .

pramodsatya commented:
Hi @aditi-pandit, could you please help review these changes? The second commit contains changes specific to query runner context.

aditi-pandit commented on velox/exec/fuzzer/ReferenceQueryRunner.h:
```diff
@@ -19,6 +19,11 @@
 
+class QueryRunnerContext {
+ public:
+  std::unordered_map<core::PlanNodeId, std::vector<std::string>> windowFrames_;
```
You could use a pair or struct with 2 fields for windowFrames_ as you are always guaranteed to have 2 of them.

aditi-pandit replied:
This is fine.

kagamiori commented on velox/exec/fuzzer/WindowFuzzer.cpp:
```diff
@@ -68,20 +69,158 @@ void WindowFuzzer::addWindowFunctionSignatures(
+    velox::test::VectorMaker vectorMaker{pool_.get()};
+
+    auto type = CppToType<T>::create();
+    VectorPtr fuzzOffset = vectorFuzzer_.fuzzConstant(type, 1);
+    offsetValue = fuzzOffset->as<ConstantVector<T>>()->valueAt(0);
+
+    while (offsetValue < 0) {
+      fuzzOffset = vectorFuzzer_.fuzzConstant(type, 1);
+      offsetValue = fuzzOffset->as<ConstantVector<T>>()->valueAt(0);
+    }
```
Instead of using a while loop, let's determine the random offset in the valid range first and then create a constant vector from it.

pramodsatya replied:
Reverted the while loop, this was added only for debugging purpose.

kagamiori commented on velox/exec/fuzzer/WindowFuzzer.cpp:
```diff
@@ -96,20 +235,42 @@ std::tuple<std::string, bool> WindowFuzzer::generateFrameClause() {
+        if (windowType == core::WindowNode::WindowType::kRange) {
+          if (windowType == core::WindowNode::WindowType::kRange) {
+            if (isPrestoQuery) {
+              return isStartBound
+                  ? fmt::format("{} PRECEDING", prestoFrames_[0].first)
+                  : fmt::format("{} PRECEDING", prestoFrames_[0].second);
```
Pass prestoFrames_[0].first or prestoFrames_[0].second to frameBound() so that this branch on isStartBound is not needed.

pramodsatya replied:
Updated, thanks.

kagamiori commented on velox/exec/fuzzer/WindowFuzzer.cpp:
```diff
@@ -235,23 +402,77 @@ void WindowFuzzer::go() {
+    auto [frameType, startBoundType, endBoundType] =
+        frameWindowTypeAndBoundType();
+
+    std::vector<core::WindowNode::BoundType> kBoundOptions = {
+        core::WindowNode::BoundType::kPreceding,
+        core::WindowNode::BoundType::kFollowing};
+    bool isKRangeFrame = frameType == core::WindowNode::WindowType::kRange &&
+        (std::find(
+             kBoundOptions.begin(), kBoundOptions.end(), startBoundType) !=
+             kBoundOptions.end() ||
+         std::find(kBoundOptions.begin(), kBoundOptions.end(), endBoundType) !=
+             kBoundOptions.end());
```
Let's extract this logic into a helper method.

kagamiori commented on velox/exec/fuzzer/WindowFuzzer.cpp:
```diff
@@ -235,23 +402,77 @@ void WindowFuzzer::go() {
+        // NOTE: velox user exception is accepted as it is caused by the invalid
+        // fuzzer test inputs. Catches arithmetic overflow exceptions in case
+        // value of orderBy column +/- offset exceeds the datatype limits.
+        stats_.numFailed++;
```
No need to increment numFailed since the query doesn't run at all. This counter is for queries that run and fail.

pramodsatya replied:
Reverted, thanks for pointing it.

kagamiori reviewed (commented):
Hi @pramodsatya, thank you for putting together this draft! I left some comments. Have you tried whether the current code already work properly with PrestoQueryRunner?

pramodsatya commented on velox/exec/fuzzer/WindowFuzzer.cpp:
```diff
@@ -235,23 +402,77 @@ void WindowFuzzer::go() {
     // If the function is order-dependent or uses "rows" frame, sort all input
     // rows by row_number additionally.
-    if (requireSortedInput || isRowsFrame) {
+    if ((requireSortedInput && !isKRangeFrame) ||
+        frameType == core::WindowNode::WindowType::kRows) {
```
Could you please share why the ties need to be broken in the order by column in case of kRange frames? This column was removed for kRange frames since we want only one orderBy column, which is randomly generated with vectorFuzzer.

pramodsatya replied:
Hi @kagamiori, thanks for pointing this out. I am planning to use the information about order dependent functions to decide what the order by column would be. When we have kRange frames: if the function is order dependent, `row_number` will be used as the order by column, otherwise the order by column will be generated randomly and `row_number` column will not be added. will work on the prototype, compare the results with Presto query runner and share the findings by today.

pramodsatya commented on velox/exec/tests/utils/QueryAssertions.cpp:
```diff
@@ -711,16 +711,16 @@ std::string MaterializedRowEpsilonComparator::getUserFriendlyDiff(
-      extraRows.push_back(actualRow);
-      ++actualIndex;
-    }
+    //    if (equalWithEpsilon(expectedRow, actualRow)) {
```
Yes this was just for debugging, it is reverted now.

pramodsatya reviewed (commented):
Thanks for the feedback @kagamiori. We found some result mismatches with Presto while testing with these changes and are investigating them further.

aditi-pandit commented on velox/exec/fuzzer/AggregationFuzzerBase.cpp:
```diff
@@ -236,14 +236,41 @@ std::vector<std::string> AggregationFuzzerBase::generateKeys(
     std::vector<std::string>& names,
-    std::vector<TypePtr>& types) {
+    std::vector<TypePtr>& types,
+    const bool hasRowNumberKey,
```
Does hasRowNumberKey need to be passed as a parameter ? Can we find it from the names vector instead ?

aditi-pandit replied:
Don't think const is required here. The underlying function cannot change the value anyways.

aditi-pandit commented on velox/exec/fuzzer/AggregationFuzzerBase.cpp:
```diff
@@ -794,11 +828,14 @@ std::unique_ptr<ReferenceQueryRunner> setupReferenceQueryRunner(
         prestoUrl,
         runnerName,
         static_cast<std::chrono::milliseconds>(reqTimeoutMs));
+    prestoQueryRunner->queryRunnerContext_ =
```
This should be part of PrestoQueryRunner constructor itself. Why is queryRunnerContext set like this ?

aditi-pandit commented on velox/exec/fuzzer/WindowFuzzer.cpp:
```diff
@@ -60,20 +61,148 @@ void WindowFuzzer::addWindowFunctionSignatures(
+      ? core::WindowNode::WindowType::kRows
+      : core::WindowNode::WindowType::kRange;
+
+  std::vector<core::WindowNode::BoundType> startBoundOptions, endBoundOptions;
```
Generally the declaration and definition are both in the same line. Separate startBoundOptions and endBoundOptions like this.
std::vector<core::WindowNode::BoundType> startBoundOptions = {
      core::WindowNode::BoundType::kUnboundedPreceding,
      core::WindowNode::BoundType::kPreceding,
      core::WindowNode::BoundType::kCurrentRow,
      core::WindowNode::BoundType::kFollowing};

aditi-pandit commented on velox/exec/fuzzer/WindowFuzzer.cpp:
```diff
@@ -60,20 +61,148 @@ void WindowFuzzer::addWindowFunctionSignatures(
+      core::WindowNode::BoundType::kFollowing,
+      core::WindowNode::BoundType::kUnboundedFollowing};
+
+  // End bound option should not be greater than start bound option as this
```
This comment is confusing because endBoundOption is always greater than start bound options.

aditi-pandit commented on velox/exec/fuzzer/WindowFuzzer.h:
```diff
@@ -131,6 +167,11 @@ class WindowFuzzer : public AggregationFuzzerBase {
     void print(size_t numIterations) const;
   } stats_;
+
+  // Represents the value of frame bound for k-range frames to be used when
```
Please give an example of what this looks like.

aditi-pandit commented on velox/exec/fuzzer/WindowFuzzer.cpp:
```diff
@@ -88,65 +217,50 @@ std::tuple<std::string, bool> WindowFuzzer::generateFrameClause() {
-      case core::WindowNode::BoundType::kPreceding:
+      case core::WindowNode::BoundType::kPreceding: {
+        if (windowType == core::WindowNode::WindowType::kRange) {
+          if (sqlFrame) {
```
Might be better to pass sqlFrame as an argument to this lambda.

aditi-pandit commented on velox/exec/fuzzer/WindowFuzzer.cpp:
```diff
@@ -227,23 +382,75 @@ void WindowFuzzer::go() {
+    bool isKRangeFrame = frameType == core::WindowNode::WindowType::kRange &&
+        (isKBoundFrame(startBoundType) || isKBoundFrame(endBoundType));
+    auto hasRowNumberKey =
+        requireSortedInput && frameType == core::WindowNode::WindowType::kRows;
```
The previous code is OR of the 2 conditions. Why did you change to & instead ?

aditi-pandit commented on velox/exec/fuzzer/WindowFuzzer.cpp:
```diff
@@ -227,23 +382,75 @@ void WindowFuzzer::go() {
+    if (isKRangeFrame) {
+      // Catch possible type overflow errors when generating offset columns.
+      try {
+        auto orderByKey = sortingKeysAndOrders.back();
```
Since this should have exactly one key, you should add a VELOX_CHECK for the size of sortingKeysAndOrders and pick up the value at the first index instead of using back().

aditi-pandit commented on velox/exec/fuzzer/WindowFuzzer.cpp:
```diff
@@ -192,6 +309,44 @@ WindowFuzzer::generateSortingKeysAndOrders(
+
+  using TCpp = typename TypeTraits<TKind>::NativeType;
+  std::string colName;
+  std::pair<std::string, std::string> prestoFrame;
```
Nit prestoFrames

aditi-pandit commented on velox/exec/fuzzer/WindowFuzzer.cpp:
```diff
@@ -60,20 +61,148 @@ void WindowFuzzer::addWindowFunctionSignatures(
+  if constexpr (!(std::is_same_v<T, bool> || std::is_same_v<T, StringView> ||
+                  std::is_same_v<T, Timestamp> ||
+                  std::is_same_v<T, UnknownValue>)) {
+    auto size = vectorFuzzer_.getOptions().vectorSize;
```
Move size closer to where it is used.

aditi-pandit commented on velox/exec/fuzzer/WindowFuzzer.cpp:
```diff
@@ -60,20 +61,148 @@ void WindowFuzzer::addWindowFunctionSignatures(
+                  std::is_same_v<T, Timestamp> ||
+                  std::is_same_v<T, UnknownValue>)) {
+    auto size = vectorFuzzer_.getOptions().vectorSize;
+    velox::test::VectorMaker vectorMaker{pool_.get()};
```
Do we really need vectorMaker for the rowVector at the end ? It seems un-necessary.

aditi-pandit commented on velox/exec/fuzzer/WindowFuzzer.cpp:
```diff
@@ -60,20 +61,148 @@ void WindowFuzzer::addWindowFunctionSignatures(
+      newNames.push_back(columnName);
+      auto newChildren = input[i]->children();
+      newChildren.push_back(offsetColumn);
+      input[i] = vectorMaker.rowVector(newNames, newChildren);
```
We can use makeRowVector here. Do we really need vectorMaker ?

aditi-pandit commented on velox/exec/fuzzer/WindowFuzzer.cpp:
```diff
@@ -60,20 +61,148 @@ void WindowFuzzer::addWindowFunctionSignatures(
+}
+
+template <typename T>
+std::optional<std::string> WindowFuzzer::addKRangeFrameColumnToInput(
```
We should ensure that we send a column of type that can be handled in this function. Returning nullOpt is weird. We are generating columns of the correct types. Just adding a VELOX_CHECK for the type here is sufficient.

aditi-pandit reviewed (commented):
Thanks @pramodsatya. This is beginning to look solid.

pramodsatya commented:
Hi @aditi-pandit, @kagamiori, could you please help review this PR?

aditi-pandit commented on velox/exec/fuzzer/WindowFuzzer.h:
```diff
@@ -124,6 +160,27 @@ class WindowFuzzer : public AggregationFuzzerBase {
+  // node would look be: std::vector{ std::pair{"2", "3"}, std::pair{"", "3"} },
+  // and the frame clause in Presto, corresponding to the two window operators
+  // would be "RANGE BETWEEN 2 PRECEDING AND 3 FOLLOWING", "RANGE BETWEEN
+  // UNBOUNDED PRECEDING AND 9 FOLLOWING". This frame clause can be constructed
```
Why not use string UNBOUNDED PRECEDING in the frame instead of "". Would be cleaner.

aditi-pandit commented on velox/exec/fuzzer/WindowFuzzer.cpp:
```diff
@@ -60,20 +61,132 @@ void WindowFuzzer::addWindowFunctionSignatures(
+    core::WindowNode::WindowType,
+    core::WindowNode::BoundType,
+    core::WindowNode::BoundType>
+WindowFuzzer::frameWindowTypeAndBoundType() {
```
generateFrameClause was a better name for this method. Is it possible to retain it ?

aditi-pandit commented on velox/exec/fuzzer/WindowFuzzer.cpp:
```diff
@@ -60,20 +61,132 @@ void WindowFuzzer::addWindowFunctionSignatures(
+  return fuzzOffset->toString(0);
+}
+
+std::string WindowFuzzer::generateFrameClause(
```
This could be called frameClauseString ?

aditi-pandit reviewed (commented):
Thanks @pramodsatya.Only few minor comments left.

pramodsatya commented:
> Thanks @pramodsatya.Only few minor comments left.

Thanks @aditi-pandit, addressed the comments.

aditi-pandit commented:
@pramodsatya : Please rebase your code. There is a conflict.

aditi-pandit reviewed (approved):
Thanks @pramodsatya

kagamiori commented on velox/exec/fuzzer/WindowFuzzer.h:
```diff
@@ -80,9 +80,42 @@ class WindowFuzzer : public AggregationFuzzerBase {
+      SortingKeyAndOrder& orderByKey);
+
+  // Add offset column to input data for k-range frames. Returns the value of K
+  // as a string.
```
In the comment,  let's also mention that this method generate a random constant k for the frame bound and update `input` with a column of the adjusted offset between k and the sorting key value.

kagamiori commented on velox/exec/fuzzer/WindowFuzzer.cpp:
```diff
@@ -60,20 +61,132 @@ void WindowFuzzer::addWindowFunctionSignatures(
+const std::string WindowFuzzer::addKRangeOffsetColumnToInput(
+    std::vector<RowVectorPtr>& input,
+    core::WindowNode::BoundType frameBoundType,
+    std::string& columnName,
+    SortingKeyAndOrder& orderByKey) {
```
This method doesn't change columnName and orderByKey, so let's pass them as const reference.

kagamiori commented on velox/exec/fuzzer/WindowFuzzer.cpp:
```diff
@@ -60,20 +61,132 @@ void WindowFuzzer::addWindowFunctionSignatures(
+    core::WindowNode::WindowType windowType,
+    core::WindowNode::BoundType startBoundType,
+    core::WindowNode::BoundType endBoundType,
+    bool sqlFrame) {
```
nit: Should it be called prestoSqlFrame since it is specific to PrestoQueryRunner?

kagamiori commented on velox/exec/fuzzer/AggregationFuzzerBase.cpp:
```diff
@@ -241,14 +241,41 @@ std::vector<std::string> AggregationFuzzerBase::generateKeys(
+  // row_number column is not present, generate this sorting key randomly; use
+  // the row_number column as sorting key otherwise.
+  if (rangeFrame) {
+    if (hasRowNumberKey) {
```
nit: useRowNumberKey sounds more accurate.

kagamiori commented on velox/exec/fuzzer/WindowFuzzer.cpp:
```diff
@@ -191,6 +292,48 @@ std::vector<SortingKeyAndOrder> WindowFuzzer::generateSortingKeysAndOrders(
 }
 
+template <TypeKind TKind>
+void WindowFuzzer::addOffsetColumnsToInput(
```
nit: Since this method is specific to k-range frames, let's rename it as addKRangeOffsetColumnToInput. The original addKRangeOffsetColumnToInput can be renamed as addKRangeOffsetColumnToInputImpl.

kagamiori commented on velox/exec/fuzzer/WindowFuzzer.cpp:
```diff
@@ -226,23 +369,80 @@ void WindowFuzzer::go() {
     // If the function is order-dependent or uses "rows" frame, sort all input
     // rows by row_number additionally.
-    if (requireSortedInput || isRowsFrame) {
+    if (hasRowNumberKey) {
       sortingKeysAndOrders.emplace_back("row_number", core::kAscNullsLast);
       ++stats_.numSortedInputs;
     }
 
+    TypeKind orderByTypeKind;
+    // kRange frames need only one order by key. This should be row_number for
+    // functions that are order dependent.
+    if (isKRangeFrame) {
+      sortingKeysAndOrders = generateSortingKeysAndOrders(
+          "s", argNames, argTypes, hasRowNumberKey, isKRangeFrame);
+      orderByTypeKind = argTypes[argTypes.size() - 1]->kind();
+    } else if (vectorFuzzer_.coinToss(0.5)) {
+      // 50% chance without order-by clause.
+      sortingKeysAndOrders =
+          generateSortingKeysAndOrders("s", argNames, argTypes);
+    }
```
These three branches can be combined for better readability. Then inside generateSortingKeysAndOrders(), we don't need to handle the case when useRowNumberKey is true.
```
TypeKind orderByTypeKind;
if (useRowNumberKey) {
  sortingKeysAndOrders.emplace_back("row_number", core::kAscNullsLast);
  ++stats_.numSortedInputs;
} else if (isKRangeFrame) {
  sortingKeysAndOrders = generateSortingKeysAndOrders(
          "s", argNames, argTypes, hasRowNumberKey, isKRangeFrame);
  orderByTypeKind = argTypes[argTypes.size() - 1]->kind();
} else if (vectorFuzzer_.coinToss(0.5)) {
  ...
}
```

kagamiori commented on velox/exec/fuzzer/WindowFuzzer.cpp:
```diff
@@ -226,23 +369,80 @@ void WindowFuzzer::go() {
+            endBoundType,
+            orderByKey);
+      } catch (VeloxUserError& e) {
+        continue;
```
Could we add a log before continuing that `fmt::format("This iteration is not valid due to exception from addKRangeOffsetColumnsToInput: {}", e.message())`?

kagamiori commented on velox/exec/fuzzer/WindowFuzzer.cpp:
```diff
@@ -226,23 +369,80 @@ void WindowFuzzer::go() {
+
+    const auto frameClause =
+        frameClauseString(frameType, startBoundType, endBoundType);
+    std::string prestoFrameClause;
+    if (isKRangeFrame && FLAGS_enable_window_reference_verification) {
+      if (auto* prestoQueryRunner =
+              dynamic_cast<PrestoQueryRunner*>(referenceQueryRunner_.get())) {
+        prestoFrameClause =
+            frameClauseString(frameType, startBoundType, endBoundType, true);
+      }
+    } else {
+      prestoFrameClause = frameClause;
+    }
```
nit: 
```
auto prestoFrameClause = frameClause;
if (isKRangeFrame && FLAGS_enable_window_reference_verification && auto* prestoQueryRunner =
              dynamic_cast<PrestoQueryRunner*>(referenceQueryRunner_.get())) {
  prestoFrameClause =
            frameClauseString(frameType, startBoundType, endBoundType, true);
}
```

kagamiori commented on velox/exec/fuzzer/WindowFuzzer.cpp:
```diff
@@ -406,6 +610,8 @@ bool WindowFuzzer::verifyWindow(
     if (!customVerification) {
       if (resultOrError.result && enableWindowVerification) {
+        referenceQueryRunner_->queryRunnerContext_->windowFrames_[windowNodeId]
+            .push_back(prestoFrameClause);
```
Do we clear referenceQueryRunner_->queryRunnerContext_->windowFrames_ after every fuzzer iteration?

kagamiori commented on velox/exec/fuzzer/WindowFuzzer.h:
```diff
@@ -80,9 +80,42 @@ class WindowFuzzer : public AggregationFuzzerBase {
+      SortingKeyAndOrder& orderByKey);
+
+  template <typename T>
+  T genOffsetAtIdx(
```
Could we add a comment about what this method does?

kagamiori reviewed (commented):
Hi @pramodsatya, thank you for adding this support and iterating on the PR! I left some comments. Have you tried running the window fuzzer locally with this extension with PrestoQueryRunner? I wonder if it could run with no error for something like 1 hour?

pramodsatya reviewed (commented):
Thanks for the review @kagamiori, addressed the comments and added changes for k-RANGE frames with column bounds. The `generateFrameClause` function is also broken down to separate the frame metadata generation from the frame clause string generation, following the [suggestion here](https://github.com/facebookincubator/velox/pull/10494#pullrequestreview-2236373531). Could I separate the frame clause function changes to a separate PR to make this change cleaner? 

The window fuzzer seems to be failing when run for an extended duration with different errors, I am investigating the failures and will address them shortly.

pramodsatya commented on velox/exec/fuzzer/AggregationFuzzerBase.cpp:
```diff
@@ -301,13 +328,17 @@ std::vector<RowVectorPtr> AggregationFuzzerBase::generateInputData(
-  types.push_back(BIGINT());
+    const CallableSignature& signature,
+    const bool hasRowNumberKey) {
+  if (hasRowNumberKey) {
```
The fuzzer failures appeared to be because the `row_number` column was removed so added it back to input, thanks!

pramodsatya commented:
> Hi @pramodsatya, thank you for adding this support and iterating on the PR! I left some comments. Have you tried running the window fuzzer locally with this extension with PrestoQueryRunner? I wonder if it could run with no error for something like 1 hour?

The window fuzzer errors are fixed and the fuzzer is running successfully for 1hr+ without any errors, using Presto query runner for verification. Could you please take another at the changes @aditi-pandit @kagamiori ?

kagamiori commented on velox/exec/fuzzer/WindowFuzzer.cpp:
```diff
@@ -205,6 +254,126 @@ std::vector<SortingKeyAndOrder> WindowFuzzer::generateSortingKeysAndOrders(
+      } else {
+        T offsetValue;
+        if (isColumnBound) {
+          offsetValue = getConstantK(vectorFuzzer_.fuzzConstant(type, 1, true));
```
Looks like this is generated in exactly the same way as `constantOffset`. Can we use constantOffset directly then?

kagamiori commented on velox/exec/fuzzer/WindowFuzzer.cpp:
```diff
@@ -89,78 +91,124 @@ std::string WindowFuzzer::generateFrameClause(
-          argNames.push_back(colName);
-          return fmt::format("{} FOLLOWING", colName);
+        std::string frameBound;
+        if (kRangeOffsetColumns.size()) {
```
nit: if (kRangeOffsetColumns.empty()) {...} else {...}

kagamiori commented on velox/exec/fuzzer/WindowFuzzer.h:
```diff
@@ -79,14 +79,88 @@ class WindowFuzzer : public AggregationFuzzerBase {
+  // generated to be of the same type as the ORDER-BY column. The offset column
+  // value for a given row is then generated by helper function genOffsetAtIdx.
+  // The constant K value is returned as a string.
+  // 2. If the frame bound is a column, VectorFuzzer is used to generate a
+  // column of the same type as the ORDER-BY key. Returns offset column name.
```
Looks like we only generate constant frame bound columns right now. Could we clarify this in the comment so that we remember to extend it later?

kagamiori reviewed (approved):
Looks good to me overall. Thank you for iterating on the code and investigating the failures!

kagamiori commented on velox/exec/fuzzer/WindowFuzzer.cpp:
```diff
@@ -205,6 +254,126 @@ std::vector<SortingKeyAndOrder> WindowFuzzer::generateSortingKeysAndOrders(
+  velox::test::VectorMaker vectorMaker{pool_.get()};
+
+  const auto getConstantK = [&](VectorPtr fuzzOffset) -> T {
+    return fuzzOffset->values()->as<T>()[0];
```
nit: fuzzOffset->as<ConstantVector<T>>()->valueAt(0);

kagamiori reviewed (commented):
There are build failures in this PR. Could you take a look?

aditi-pandit commented on velox/functions/prestosql/fuzzer/ApproxPercentileResultVerifier.h:
```diff
@@ -473,6 +460,15 @@ class ApproxPercentileResultVerifier : public ResultVerifier {
         fmt::format("{} as w", weighted ? weightField.value() : "1::bigint"));
+    // Names of columns, which could be offset columns in case of kRange frames,
+    // or frame bounds, should be projected.
+    static const std::unordered_set<std::string> kColumnNamesKBound = {
```
Nit : Rename kKBoundColumnNames

pramodsatya reviewed (commented):
Thanks for the feedback @kagamiori @aditi-pandit, addressed the comments and the CI error is fixed now. Could you please take another look?

(Update: the Presto Fuzzer CI failure is not related to this PR, I'm seeing the same error on other PRs as well)

kagamiori commented on velox/exec/fuzzer/WindowFuzzer.cpp:
```diff
@@ -205,6 +257,145 @@ std::vector<SortingKeyAndOrder> WindowFuzzer::generateSortingKeysAndOrders(
+  VectorFuzzer::Options options = vectorFuzzer_.getOptions();
+  auto nullRatio = options.nullRatio;
+  options.nullRatio = 0;
+  vectorFuzzer_.setOptions(options);
```
nit: You can use ScopedVarSetter here so that you don't need to reset the option later.

kagamiori commented on velox/exec/fuzzer/WindowFuzzer.cpp:
```diff
@@ -205,6 +257,145 @@ std::vector<SortingKeyAndOrder> WindowFuzzer::generateSortingKeysAndOrders(
+template <typename T>
+const std::string WindowFuzzer::addKRangeOffsetColumnToInputImpl(
+    std::vector<RowVectorPtr>& input,
+    FrameMetadata& frameMetadata,
```
nit: frameMetadata is not used?

kagamiori reviewed (changes requested):
Hi @pramodsatya, with the newest change, there are some functions that take both `frame` and `frameClause` (or similar variable names) as arguments. I found it a bit hard to follow the distinction between these arguments. Could you add an example with what the real `frame` and `frameClause` look like for a small query and explain why both of one of them is needed? Thanks!

pramodsatya commented on velox/functions/prestosql/fuzzer/ApproxDistinctResultVerifier.h:
```diff
@@ -67,6 +67,7 @@ class ApproxDistinctResultVerifier : public ResultVerifier {
       const std::vector<SortingKeyAndOrder>& /*sortingKeysAndOrders*/,
       const core::WindowNode::Function& function,
       const std::string& frame,
+      const std::string& /*frameClause*/,
```
Reverted this change and using the existing frame argument, thanks.

pramodsatya reviewed (commented):
Hi @kagamiori, the argument `frame` used to contain the complete window frame clause including the `partition by`, `order by`, and frame clauses. `frameClause` argument was added to reuse the existing variable representing just the frame clause. It is not needed in this PR though and reverted this change, apologies for the mixup. Could you please take another look?

kagamiori commented on velox/exec/fuzzer/WindowFuzzer.cpp:
```diff
@@ -89,78 +90,128 @@ std::string WindowFuzzer::generateFrameClause(
+      return checkedPlus<T>(orderByValue, frameBound);
+    }
+  }
+  return std::nullopt;
```
@pramodsatya Could you `return T{}` after VELOX_UNREACHABLE to suppress that no-value-returned error? I think putting VELOX_UNREACHABLE in this method makes it clearer that this is a faulty branch that should not be taken.

kagamiori commented on velox/exec/fuzzer/WindowFuzzer.cpp:
```diff
@@ -89,78 +90,128 @@ std::string WindowFuzzer::generateFrameClause(
+      }
       default:
-        VELOX_UNREACHABLE("Unknown option for frame clause generation");
+        return "";
```
Ditto.

kagamiori commented on velox/exec/fuzzer/PrestoQueryRunner.cpp:
```diff
@@ -446,7 +447,12 @@ std::optional<std::string> PrestoQueryRunner::toSql(
     }
 
-    appendWindowFrame(functions[i].frame, sql);
+    if (queryRunnerContext_->windowFrames_.find(windowNode->id()) ==
+        queryRunnerContext_->windowFrames_.end()) {
+      appendWindowFrame(functions[i].frame, sql);
+    } else {
+      sql << queryRunnerContext_->windowFrames_.at(windowNode->id()).back();
+    }
```
@pramodsatya Could you point me to which unit tests fails? Do they fail because of changes made in this PR?

pramodsatya replied:
Hi @kagamiori, the failing test is [`toSql` in `PrestoQueryRunnerTest.cpp`](https://github.com/facebookincubator/velox/blob/94dcf028c12a892b753946563edcc4589f69fe79/velox/exec/tests/PrestoQueryRunnerTest.cpp#L179). It fails with the following error:
```
unknown file: Failure
C++ exception with description "unordered_map::at: key not found" thrown in the test body.
```
and it seems to be because the `windowNodeId` is not captured in the map `queryRunnerContext_->windowFrames_`. Before, the `PrestoQueryRunner` was always getting the frame clause from `queryRunnerContext_`.

kagamiori replied:
Hi @pramodsatya, this unit test is for the test coverage of PrestoQueryRunner::toSql(). I don't think PrestoQueryRunner::toSql(windowNode) is used anywhere other than the WindowFuzzer. Could you provide this failing test case with queryRunnerContext_->windowFrames_ and move it to a separate unit test for QueryRunnerContext in PrestoQueryRunnerTest.cpp? Thanks!

pramodsatya replied:
Updated accordingly, thanks!

kagamiori commented on velox/functions/prestosql/fuzzer/ApproxPercentileResultVerifier.h:
```diff
@@ -88,7 +89,7 @@ class ApproxPercentileResultVerifier : public ResultVerifier {
         weightField,
         sortingKeysAndOrders,
-        function.frame,
+        frameClause,
```
What's the difference between function.frame and frameClause? Also, why does ApproxPercentileResultVerifier use frameClause but ApproxDistinctResultVerifier uses frame? Is there a difference between the two verifiers that causes this distinction?

pramodsatya replied:
In `ApproxPercentileResultVerifier`, the frame clause is constructed using `function.frame` with the helper function:
```
std::string getFrameClause(const core::WindowNode::Frame& frame);
```
This function was removed as it doesn't account for frames with columns as bounds, and resulted in the error shared before. Also since `frameClause` is now passed as an argument to `verifyWindow`, it could be passed directly to `initializeWindow` in the verifier. The usage in `ApproxPercentileResultVerifier`:
```
.window({fmt::format(
            "multimap_agg(x, w) over ({} {} {}) as bucket",
            getPartitionByClause(groupingKeys_),
            getOrderByClause(sortingKeysAndOrders),
            frameClause)})
```

In `ApproxDistinctResultVerifier` on the other hand, the `over clause` in the window is used, this is passed in the argument `frame`:
```
auto plan = PlanBuilder()
                    .values(input)
                    .window({makeCountDistinctWindowCall(function, frame)})
                    .planNode();
```
Please let me know if the variable names should be changed or other refactor is needed.

kagamiori replied:
Hi @pramodsatya, Got it. Thank you for the detailed explanation. I found the reason why getFrameClause() doesn't work for column bounds: getFrameClause() uses WindowNode::Frame::startValue->toString() while startValue is a FieldAccessTypedExpr when it's a column bound. FieldAccessTypedExpr::toString() returns `ROW["off1"]` instead of the `off1` we want.

```
.window({fmt::format(
            "multimap_agg(x, w) over ({} {} {}) as bucket",
            getPartitionByClause(groupingKeys_),
            getOrderByClause(sortingKeysAndOrders),
            frameClause)})
```
This fmt::format() actually generates the same string as the `frame` argument of initializeWindow(). An easy solution is to pass this `frame` to computePercentilesForWindow(), then replace the statement above by the following.
```
.window({fmt::format(
            "multimap_agg(x, w) over ({}) as bucket",
            frame)})
```

Then you don't need to pass an additional argument `frameClause` to initializeWindow() and can remove the getPartitionByClause() and getOrderByClause(). In general, we'd like to avoid passing multiple arguments with redundant information because that could cause confusion.

pramodsatya replied:
Thanks for the feedback @kagamiori. I wasn't sure if the fmt::format output was same as `frame`, updated to use the existing argument as suggested.

kagamiori commented on velox/functions/prestosql/fuzzer/ApproxPercentileResultVerifier.h:
```diff
@@ -484,7 +480,7 @@ class ApproxPercentileResultVerifier : public ResultVerifier {
             getPartitionByClause(groupingKeys_),
             getOrderByClause(sortingKeysAndOrders),
-            getFrameClause(frame))})
+            frameClause)})
```
Hi @pramodsatya, could you give me an example where getFrameClause(frame) is different from frameClause and is wrong in this use case? Thanks!

kagamiori reviewed (commented):
Hi @pramodsatya, I saw there is a window fuzzer failure in CI. Could you please take a look? https://github.com/facebookincubator/velox/actions/runs/10730400789/job/29760243003?pr=10006

kagamiori commented on velox/functions/prestosql/fuzzer/ApproxPercentileResultVerifier.h:
```diff
@@ -303,20 +303,6 @@ class ApproxPercentileResultVerifier : public ResultVerifier {
-    return ss.str();
-  }
-
   std::string getOrderByClause(
```
nit: getOrderByClause is not used either and can be removed.

pramodsatya commented:
> Hi @pramodsatya, I saw there is a window fuzzer failure in CI. Could you please take a look? https://github.com/facebookincubator/velox/actions/runs/10730400789/job/29760243003?pr=10006

Hi @kagamiori @aditi-pandit, it looks like the failure was because of the check for null values in the `frameColumn` in function `WindowPartition::updateKRangeFrameBounds`. The order by column index obtained from `sortKeyInfo_`, in the variable `orderByColumn`, already contains the order by column index in the reordered data (since `WindowBuild` constructs `std::unique_ptr<RowContainer> data_;` by placing `partitionKeys` and `orderByKeys` first). So the `orderByRowColumn` should be obtained from the inverse map of `inputMapping_`, so the right index (which maps to `sortKeyInfo_.first` in the inverse map) is queried in `std::vector<exec::RowColumn> columns_;` for the null check.

Eg: Input
```
-- Window[1][partition by [p0, p1] order by [s0 ASC NULLS LAST] w0 := skewness(ROW["c0"]) RANGE between UNBOUNDED PRECEDING and off1 PRECEDING] -> c0:BIGINT, p0:ROW<f0:REAL,f1:INTERVAL DAY TO SECOND,f2:REAL,f3:DATE,f4:INTERVAL DAY TO SECOND,f5:VARCHAR>, p1:ARRAY<BIGINT>, s0:HUGEINT, row_number:BIGINT, off1:HUGEINT, w0:DOUBLE
  -- Values[0][1000 rows in 10 vectors] -> c0:BIGINT, p0:ROW<f0:REAL,f1:INTERVAL DAY TO SECOND,f2:REAL,f3:DATE,f4:INTERVAL DAY TO SECOND,f5:VARCHAR>, p1:ARRAY<BIGINT>, s0:HUGEINT, row_number:BIGINT, off1:HUGEINT
```
Issue
```
column_index_t orderByColumn = sortKeyInfo_[0].first;    ==> 2
inputMapping_ : 3 0 1 2 4 5
inputMapping_[orderByColumn] = 1
RowColumn corresponding to orderByColumn is at index 3 in columns_
```

Could you please take another look?

kagamiori commented:
Hi @pramodsatya, is that failure still reproducible? If so, could you share which command you used to reproduce it? Thanks!

pramodsatya commented:
> Hi @pramodsatya, is that failure still reproducible? If so, could you share which command you used to reproduce it? Thanks!

Hi @kagamiori, the failure is not reproducible currently. It can be reproduced by reverting the changes to `WindowPartition.cpp` in this PR, and with the following command:
```
velox_window_fuzzer_test --enable_window_reference_verification --presto_url="http://127.0.0.1:8080" --duration_sec=3600 --logtostderr=1 --minloglevel=0 --seed=2483741532 --req_timeout_ms=5000
```

kagamiori commented on velox/exec/WindowPartition.cpp:
```diff
@@ -364,7 +364,15 @@ void WindowPartition::updateKRangeFrameBounds(
   vector_size_t end;
   RowColumn frameRowColumn = columns_[frameColumn];
-  RowColumn orderByRowColumn = columns_[inputMapping_[orderByColumn]];
+  column_index_t mappedOrderByColumn = -1;
+  for (auto i = 0; i < inputMapping_.size(); i++) {
+    if (inputMapping_[i] == orderByColumn) {
+      mappedOrderByColumn = i;
+      break;
+    }
+  }
+  VELOX_DCHECK_NE(mappedOrderByColumn, -1);
+  RowColumn orderByRowColumn = columns_[mappedOrderByColumn];
```
So the problem with the original code is that orderByColumn is the column index into data_ where columns are already reordered to start with partition-by keys and sorting keys. On the other hand, columns_ keeps the original column order (i.e., same as the original input to the Window operator). So there was a mismatch. To make it worse, the original code accesses columns_ through `columns_[inputMapping_[orderByColumn]]` where inputMapping_ is a mapping from the original input column index to the column index after reordering (i.e., it is the reverse of what the code intended to do). 

Since columns_ was derived from data_ (see WindowPartition constructor), and both data_ and orderByColumn follow the column index after column reordering, I wonder whether the fix can be simplified to the following, so that you don't need to loop over inputMapping_ to find the original column index of orderByColumn.
```
RowColumn orderByRowColumn = data_->columnAt(orderByColumn);
```

kagamiori reviewed (commented):
Hi @pramodsatya, thank you for investigating the fuzzer failure! I can reproduce that bug, and as you explained, it was due to an incorrect use of inputMapping_ in WindowPartition::updateKRangeFrameBounds(). It's a great demonstration that this fuzzer enhancement enables us to catch a real bug.

Since this PR is already big, I wonder if we could separate the bug fix into another PR and add a unit test with that?

pramodsatya commented:
> Hi @pramodsatya, thank you for investigating the fuzzer failure! I can reproduce that bug, and as you explained, it was due to an incorrect use of inputMapping_ in WindowPartition::updateKRangeFrameBounds(). It's a great demonstration that this fuzzer enhancement enables us to catch a real bug.
> 
> Since this PR is already big, I wonder if we could separate the bug fix into another PR and add a unit test with that?

Thanks for the feedback @kagamiori, addressed the comments and opened another PR for the null check fix: https://github.com/facebookincubator/velox/pull/11075 . Could you please take another look?

pramodsatya commented:
Hi @kagamiori, after the rebase, window fuzzer is failing with kRange frames when alternative plans are tested, I'm looking into it:
```
-- Window[2][STREAMING partition by [p0, p1] order by [s0 DESC NULLS FIRST] w0 := checksum(ROW["c0"]) RANGE between off0 FOLLOWING and off1 FOLLOWING] -> c0:INTERVAL DAY TO SECOND, p0:INTERVAL DAY TO SECOND, p1:BOOLEAN, s0:REAL, row_number:BIGINT, off0:REAL, k1:REAL, off1:REAL, w0:VARBINARY
  -- OrderBy[1][p0 ASC NULLS FIRST, p1 ASC NULLS FIRST, s0 DESC NULLS FIRST] -> c0:INTERVAL DAY TO SECOND, p0:INTERVAL DAY TO SECOND, p1:BOOLEAN, s0:REAL, row_number:BIGINT, off0:REAL, k1:REAL, off1:REAL
    -- Values[0][1000 rows in 10 vectors] -> c0:INTERVAL DAY TO SECOND, p0:INTERVAL DAY TO SECOND, p1:BOOLEAN, s0:REAL, row_number:BIGINT, off0:REAL, k1:REAL, off1:REAL

Expected 1000, got 1000
1 extra rows, 1 missing rows
1 of extra rows:
        2933652544787757131 | 4798976694194643045 | true | "Infinity" | 420 | "Infinity" | "NaN" | "NaN" | "Zw90+g6vltk="

1 of missing rows:
        2933652544787757131 | 4798976694194643045 | true | "Infinity" | 420 | "Infinity" | "NaN" | "NaN" | null
```

kagamiori commented on velox/vector/fuzzer/VectorFuzzer.h:
```diff
@@ -166,7 +183,7 @@ class VectorFuzzer {
   }
 
-  const VectorFuzzer::Options& getOptions() {
+  VectorFuzzer::Options& getOptions() {
```
nit: Let's keep this returning `const VectorFuzzer::Options` and add another interface of `VectorFuzzer::Options& getMultableOptions()` to make the potential side-effect explicit.

kagamiori reviewed (commented):
LGTM. Thank you for adding this support, iterating on this PR, and fixing the fuzzer-found bug!

pramodsatya commented:
> Hi @kagamiori, after the rebase, window fuzzer is failing with kRange frames when alternative plans are tested, I'm looking into it:
> 
> ```
> -- Window[2][STREAMING partition by [p0, p1] order by [s0 DESC NULLS FIRST] w0 := checksum(ROW["c0"]) RANGE between off0 FOLLOWING and off1 FOLLOWING] -> c0:INTERVAL DAY TO SECOND, p0:INTERVAL DAY TO SECOND, p1:BOOLEAN, s0:REAL, row_number:BIGINT, off0:REAL, k1:REAL, off1:REAL, w0:VARBINARY
>   -- OrderBy[1][p0 ASC NULLS FIRST, p1 ASC NULLS FIRST, s0 DESC NULLS FIRST] -> c0:INTERVAL DAY TO SECOND, p0:INTERVAL DAY TO SECOND, p1:BOOLEAN, s0:REAL, row_number:BIGINT, off0:REAL, k1:REAL, off1:REAL
>     -- Values[0][1000 rows in 10 vectors] -> c0:INTERVAL DAY TO SECOND, p0:INTERVAL DAY TO SECOND, p1:BOOLEAN, s0:REAL, row_number:BIGINT, off0:REAL, k1:REAL, off1:REAL
> 
> Expected 1000, got 1000
> 1 extra rows, 1 missing rows
> 1 of extra rows:
>         2933652544787757131 | 4798976694194643045 | true | "Infinity" | 420 | "Infinity" | "NaN" | "NaN" | "Zw90+g6vltk="
> 
> 1 of missing rows:
>         2933652544787757131 | 4798976694194643045 | true | "Infinity" | 420 | "Infinity" | "NaN" | "NaN" | null
> ```

Hi @kagamiori, we investigated this error and it seems to be because of the [recently added changes](https://github.com/facebookincubator/velox/commit/8ad7c0c040283b53d18f1c1e553b5f2aa36f3a0a) which enable the vector fuzzer to generate Nan and Infinity values. Nan and Infinity values are now accounted for when constructing the frame offset column and the error is fixed now. 
Could you please help review this fix? will squash to a single commit if the change looks good. Thanks!
cc: @minhancao

kagamiori commented on velox/exec/fuzzer/WindowFuzzer.cpp:
```diff
@@ -286,15 +289,28 @@ VectorPtr WindowFuzzer::buildKRangeColumn(
     } else {
+      const auto orderByValueAtIdx = decodedVector.valueAt<T>(j);
+      if constexpr (std::is_same_v<T, float>) {
+        if (orderByValueAtIdx == kInfFloat ||
+            orderByValueAtIdx == std::nanf("")) {
```
nit: use std::isfinite() and std::isnan(), then you don't need to define kInfFloat and kInfDouble, and can catch nan other than std::nan("") or std::nanf("").

kagamiori reviewed (commented):
Hi @pramodsatya, I left one comment. The other part of this fix looks good to me. Thanks!

pramodsatya commented:
> Hi @pramodsatya, I left one comment. The other part of this fix looks good to me. Thanks!

Thanks for the suggestion @kagamiori, updated accordingly. Could you please help merge this change?

kagamiori commented:
Hi @pramodsatya, FYI, there was an internal linter complaint about defining a static variable in the header file VectorFuzzer.h. So I reverted the change in VectorFuzzer.h and VectorFuzzer.cpp, but instead declare the function `const std::vector<TypePtr>& defaultScalarTypes()` in VectorFuzzer.h so that AggregationFuzzerBase.cpp can use defaultScalarTypes().

I'm going to start the merging process now.

pramodsatya commented:
> Hi @pramodsatya, FYI, there was an internal linter complaint about defining a static variable in the header file VectorFuzzer.h. So I reverted the change in VectorFuzzer.h and VectorFuzzer.cpp, but instead declare the function `const std::vector<TypePtr>& defaultScalarTypes()` in VectorFuzzer.h so that AggregationFuzzerBase.cpp can use defaultScalarTypes(). I have checked there was no build issue on my Mac and Linux.
> 
> I'm going to start the merging process now.

Thanks for the update @kagamiori, sounds good. Please let me know if any other changes are needed.

Yuhta commented:
@pramodsatya Window fuzzer has been broken since this change, can you take a look?
https://github.com/facebookincubator/velox/actions/runs/11226206459/job/31207444517

kagamiori commented:
> @pramodsatya Window fuzzer has been broken since this change, can you take a look? https://github.com/facebookincubator/velox/actions/runs/11226206459/job/31207444517

@Yuhta, I'm taking a look at this failure: https://github.com/facebookincubator/velox/issues/11213.
